From a1465cad59be76794bdf43289ef3f92e9dfa40b9 Mon Sep 17 00:00:00 2001
From: chenyanxzhu <chenyanx.zhu@intel.com>
Date: Thu, 10 Nov 2022 13:28:31 +0800
Subject: [PATCH] Fix virtio2d color r/b reversed issue

Change default virtio dumb format from bgrx to rgbx, which is more
consistent with framework code. Need relevant qemu code changing
to support accept rgbx.

Tracked-On: OAM-103582
Signed-off-by: chenyanxzhu <chenyanx.zhu@intel.com>
---
 drivers/gpu/drm/drm_fourcc.c         | 2 ++
 drivers/gpu/drm/virtio/virtgpu_gem.c | 2 +-
 include/drm/drm_fourcc.h             | 4 ++++
 3 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/drm_fourcc.c b/drivers/gpu/drm/drm_fourcc.c
index eda832f9200d..77c73d7e2cf6 100644
--- a/drivers/gpu/drm/drm_fourcc.c
+++ b/drivers/gpu/drm/drm_fourcc.c
@@ -115,6 +115,8 @@ uint32_t drm_driver_legacy_fb_format(struct drm_device *dev,
 			fmt = DRM_FORMAT_HOST_RGB565;
 		if (fmt == DRM_FORMAT_XRGB1555)
 			fmt = DRM_FORMAT_HOST_XRGB1555;
+		if (fmt == DRM_FORMAT_XBGR8888)
+			fmt = DRM_FORMAT_HOST_XBGR8888;
 	}
 
 	if (dev->mode_config.quirk_addfb_prefer_xbgr_30bpp &&
diff --git a/drivers/gpu/drm/virtio/virtgpu_gem.c b/drivers/gpu/drm/virtio/virtgpu_gem.c
index 2de61b63ef91..30f61c6046ed 100644
--- a/drivers/gpu/drm/virtio/virtgpu_gem.c
+++ b/drivers/gpu/drm/virtio/virtgpu_gem.c
@@ -75,7 +75,7 @@ int virtio_gpu_mode_dumb_create(struct drm_file *file_priv,
 	args->size = pitch * args->height;
 	args->size = ALIGN(args->size, PAGE_SIZE);
 
-	params.format = virtio_gpu_translate_format(DRM_FORMAT_HOST_XRGB8888);
+	params.format = virtio_gpu_translate_format(DRM_FORMAT_HOST_XBGR8888);
 	params.width = args->width;
 	params.height = args->height;
 	params.size = args->size;
diff --git a/include/drm/drm_fourcc.h b/include/drm/drm_fourcc.h
index 3b138d4ae67e..72a96bb75e48 100644
--- a/include/drm/drm_fourcc.h
+++ b/include/drm/drm_fourcc.h
@@ -40,11 +40,15 @@
 				       DRM_FORMAT_BIG_ENDIAN)
 # define DRM_FORMAT_HOST_XRGB8888     DRM_FORMAT_BGRX8888
 # define DRM_FORMAT_HOST_ARGB8888     DRM_FORMAT_BGRA8888
+# define DRM_FORMAT_HOST_XBGR8888     DRM_FORMAT_RGBX8888
+# define DRM_FORMAT_HOST_ABGR8888     DRM_FORMAT_RGBA8888
 #else
 # define DRM_FORMAT_HOST_XRGB1555     DRM_FORMAT_XRGB1555
 # define DRM_FORMAT_HOST_RGB565       DRM_FORMAT_RGB565
 # define DRM_FORMAT_HOST_XRGB8888     DRM_FORMAT_XRGB8888
 # define DRM_FORMAT_HOST_ARGB8888     DRM_FORMAT_ARGB8888
+# define DRM_FORMAT_HOST_XBGR8888     DRM_FORMAT_XBGR8888
+# define DRM_FORMAT_HOST_ABGR8888     DRM_FORMAT_ABGR8888
 #endif
 
 struct drm_device;
-- 
2.38.1

